### PR TITLE
Fix a crash by removing the RestateFilteringConsumer

### PR DESCRIPTION
**Pull request by slavapestov**
The implementation of RestateFilteringConsumer was building a new GenericSignature "from scratch" by calling `GenericSignature::get()`. This violates GenericSignatureBuilder invariants and could cause crashes.

Instead, note that OverrideFilteringConsumer can already do most of what RestateFilteringConsumer did. Clean up and refactor things until RestateFilteringConsumer is removed altogether.

Fixes rdar://problem/46853611.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
Nice! I'm wondering this fixes duplicated completion issues such as: rdar://problem/21106642, rdar://problem/35992916, https://bugs.swift.org/browse/SR-1181

e.g.
```swift
protocol P {}
extension P {
  func foo() {}
}
struct S : P {
  func foo() {}
}

func test(s: S) {
  s.#^COMPLETE^#
}
```

**Comment by slavapestov** [comment]
@rintaro It might, I’ll look at those bugs in a bit.

File: lib/Sema/LookupVisibleDecls.cpp [hunk 1]
```diff
@@ -801,7 +699,9 @@ class OverrideFilteringConsumer : public VisibleDeclConsumer {
 
     // If this kind of declaration doesn't participate in overriding, there's
     // no filtering to do here.
-    if (!isa<AbstractFunctionDecl>(VD) && !isa<AbstractStorageDecl>(VD)) {
+    if (!isa<AbstractFunctionDecl>(VD) &&
+        !isa<AbstractStorageDecl>(VD) &&
+        !isa<AssociatedTypeDecl>(VD)) {
```
Comment (DougGregor) [inline lib/Sema/LookupVisibleDecls.cpp]: Technically, associated types do have "overrides" (when an associated type is restated in an inheriting protocol).

**Review by DougGregor — COMMENTED** [review]
This looks fairly safe, despite its size.

File: lib/Sema/LookupVisibleDecls.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/Sema/LookupVisibleDecls.cpp]: Yeah, notice the 'if' check is an early return, so we now handle associated types whereas before we skipped them.

Previously the RestatedFilteringConsumer would catch restated associated types, but now it has to be handled in the OverrideFilteringConsumer.

**Comment by slavapestov** [comment]
@rintaro So I looked at those bugs and a few related radars. Some were already fixed on master. The ones that were not fixed were not addressed by this PR, which is OK since the goal was to fix a crash. :) I have another PR coming up with a few more LookupVisibleDecls fixes.

**Comment by rintaro** [comment]
Thank you!

**Pull request merged**